### PR TITLE
Refactor chip class and move independent chips into separate classes

### LINKED ISSUE #101504: Refactor `chip.dart`: Move independent chips into separate classes
Currently, independent chips like `ActionChip`, `ChoiceChip`, `FilterChip`, and `InputChip` are part of `chip.dart` and the new M3 migration will add two new chips with new names,` chip.dart` is already bloated.

### Solution

Refactor `chip.dart` and move independent chips into separate classes

Context: https://github.com/flutter/flutter/issues/99024#issuecomment-1091560959

### PR DESCRIPTION
**Pull request by TahaTesser**
Currently, independent chips like `ActionChip`, `ChoiceChip`, `FilterChip`, and `InputChip` are part of `chip.dart` and the new M3 migration will add two new chips with new names,` chip.dart` is already bloated.

fixes https://github.com/flutter/flutter/issues/101504

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/test/material/chip_action_test.dart [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        ),
+      ),
+    );
+
+    await tester.tap(find.byType(ActionChip));
+    expect(tester.takeException(), null);
+  });
+
+  void checkChipMaterialClipBehavior(WidgetTester tester, Clip clipBehavior) {
+    final Iterable<Material> materials = tester.widgetList<Material>(find.byType(Material));
```
Comment (HansMuller) [inline packages/flutter/test/material/chip_action_test.dart:45]: Should explain why we expect two Material widgets here and why the last one is the one we want.

File: packages/flutter/test/material/chip_choice_test.dart [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+  return tester.widget(
+    find.ancestor(
+      of: find.text(labelText),
+      matching: find.byType(DefaultTextStyle),
+    ).first,
+  );
+}
+
+/// Adds the basic requirements for a Chip.
+Widget _wrapForChip({
```
Comment (HansMuller) [inline packages/flutter/test/material/chip_choice_test.dart:38]: If all of the other helper functions are public, then this one probably should be too...

File: packages/flutter/test/material/chip_filter_test.dart [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+import '../rendering/mock_canvas.dart';
+
+/// Adds the basic requirements for a Chip.
+Widget _wrapForChip({
```
Comment (HansMuller) [inline packages/flutter/test/material/chip_filter_test.dart:11]: Same comment as with previous tests: should these all be public?

**Review by HansMuller — APPROVED** [review]
I realize that you're just moving classes into new files here, so my comments about being consistent wrt to public/private helper methods just applies to the pre-existing code.
a
This looks like a healthy refactor.

**Comment by HansMuller** [comment]
@guidezpl - The "Google testing" failure is:
```
google3:///[mobile/flutter/tools/icon_font/bin/icon_map_generator.dart:46](https://cs.corp.google.com/piper///depot/google3/mobile/flutter/tools/icon_font/bin/icon_map_generator.dart?l=46&ws=tap-presubmit-rerun-server/625852&snapshot=2):22: Error: Too few positional arguments: 2 required, 1 given.
    final icon = Icon(entry);
                     ^
google3:///[third_party/dart/flutter/tools/lib/update_icons.dart:393](https://cs.corp.google.com/piper///depot/google3/third_party/dart/flutter/tools/lib/update_icons.dart?l=393&ws=tap-presubmit-rerun-server/625852&snapshot=2):3: Context: Found this candidate, but the arguments don't match.
  Icon(MapEntry<String, String> tokenPair, this.fontFamily) {
  ^^^^
```
Presumably this has already been fixed?

ERROR: mobile/flutter/tools/icon_font/bin/icon_map_generator.dart

46: final icon = Icon(entry); 2 positional argument(s) expected, but 1 found. #not_enough_positional_arguments

**Comment by HansMuller** [comment]
@TahaTesser - I'm guessing that what needs to be done here is to rebase and push another commit. The Google testing failure doesn't appear to have anything to do with this PR.

**Comment by TahaTesser** [comment]
@HansMuller 

I made other private helper functions public since most of them are public even in other classes and add the comments to the material helper function :) 

Yes, Google testing has been affecting many PRs lately, I will rebase now. 

Some PRs just get stuck at Google testing such as 
https://github.com/flutter/flutter/pull/101345
https://github.com/flutter/flutter/pull/101512

**Comment by guidezpl** [comment]
Yeah this was fixed internally, so all PRs need to rebase.

**Pull request merged**